### PR TITLE
Demo for 2D-IR-Raman spectrum calculation using equilibrium-nonequilibrium RPMD

### PR DESCRIPTION
**Pull request by tbegusic**
Includes a demo for running equilibrium-nonequilibrium RPMD with additional scripts in demos/2D-IR-Raman and a Fortran driver that computes induced dipole moment and polarizability of liquid water. @litman90

### CODE REVIEW

**Comment by litman90** [comment]
Thanks Tomislav, I will go over it in the following days

File: demos/2D-IR-Raman/noneqm-traj.py [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+        help="Define the verbosity level.",
+    )
+
+    parser.add_option(
+        "-e",
+        "--epsilon",
+        dest="epsilon",
+        type="float",
+        default=0.1,
+        help="Epsilon parameter controlling the field strength.",
```
Comment (litman90) [inline demos/2D-IR-Raman/noneqm-traj.py:186]: Is this atomic units? Please specify

File: demos/2D-IR-Raman/README.md [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+```
+To tell i-PI to use both lammps (for forces) and f90 driver (for properties), we need to add both of these as ffsockets and "forces" (note: water_dip_pol model returns zero force, so it does not affect dynamics). In run.sh, the equilibrium trajectory is then run using:
+```bash
+i-pi input.xml &> output &
+sleep 10
+$lmp < in.lmp &> out_lammps &
+i-pi-driver -u -h h2o-dipole -m water_dip_pol -o 1 &> out
+wait
+```
+where $lmp is the lammps executable. Note that i-pi-driver option water_dip_pol accepts one argument that is 1 if we want to compute dipole gradient or 0 if not.
```
Comment (litman90) [inline demos/2D-IR-Raman/README.md:66]: I think dipole gradient here means dipole-z gradient, right? Please specify

File: demos/2D-IR-Raman/noneqm-traj.py [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+                sim.run()
+                #############################################################################################
+                # Stop the thread that monitors for softexit. This is needed because a new thread is started
+                # every time we invoke simulation run, leading to a constant increase in the number of
+                # threads. The code breaks when a maximum number of threads is reached.
+                softexit._doloop[0] = False
+                while softexit._thread.is_alive():
+                    softexit._thread.join(0.5)
+                #############################################################################################
+
```
Comment (litman90) [inline demos/2D-IR-Raman/noneqm-traj.py:140]: Add here a job finished (without errors?) message

File: demos/2D-IR-Raman/noneqm-response.py [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+                c += 1
+        r_t *= self.beta / (c * self.epsilon)
+        r_t = np.gradient(r_t, self.step_print * self.delta_t, axis=0, edge_order=2)
+
+        np.savetxt(
+            self.out_name + self.op[0] + "_" + self.op[1] + "_neq_2d.dat",
+            r_t,
+            fmt="%.5e",
+        )
+
```
Comment (litman90) [inline demos/2D-IR-Raman/noneqm-response.py:158]: Please add an exit message that species the name of the output file created

File: demos/2D-IR-Raman/noneqm-response.py [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+
+if __name__ == "__main__":
+    from optparse import OptionParser
+
+    parser = OptionParser(
+        usage="%prog [options] <input file>",
+        description="noneqm-response comutes response function of "
+        "2D IR-Raman spectra from equilibrium and nonequilibrium trajectories",
+    )
```
Comment (litman90) [inline demos/2D-IR-Raman/noneqm-response.py:167]: Can you also explain here the output format, i.e. what are the rows and the columns?

File: demos/2D-IR-Raman/noneqm-response.py [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+                # Compute r_t[j, k] = eq[j] * neq[k].
+                r_t += np.outer(eq, neq)
+                c += 1
+        r_t *= self.beta / (c * self.epsilon)
+        r_t = np.gradient(r_t, self.step_print * self.delta_t, axis=0, edge_order=2)
+
+        np.savetxt(
+            self.out_name + self.op[0] + "_" + self.op[1] + "_neq_2d.dat",
+            r_t,
+            fmt="%.5e",
```
Comment (litman90) [inline demos/2D-IR-Raman/noneqm-response.py:156]: Please add header as comment line in the file

**Review by litman90 — CHANGES_REQUESTED** [review]
Everything is quite clear and easy to reproduce. 
I left some minor comments, please address them at your earliest convenience, add a short description to demos/README.md, and we will be ready to merge

**Comment by tbegusic** [comment]
Hi Yair, I believe I addressed all your comments. Please check and let me know if I missed something. Thank you!

**Review by litman90 — APPROVED** [review]

**Review by litman90 — APPROVED** [review]

**Comment by litman90** [comment]
Perfect, everything looks great.
I think I solved the conflict. If all the test past I will merge.
Thanks @tbegusic !

**Pull request merged**